mipengine/algorithms/metrics.py

```diff
@@ -0,0 +1,187 @@
+import numpy
```

Comment:
If the visualization computations are dependent, serial execution is unavoidable. If not, as it is the case here, a clean and generic way to approach this problem is by having an async run udf API, as we discussed. This will not give the best possible speedup but will certainly improve execution time. There are other approaches with better result but much more complicated, like writing an optimizer. 

Until then we can merge rounds as you say, but this leads to poor code and poor maintainability. The tradeoff is OK in other cases where the speedup is important but not in this particular case, because the algorithm, anyway, needs a lot of rounds, and removing one won't have such an impact.